torch/distributed/rpc/rref_proxy.py

```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
 def _local_invoke_async_execution(rref, func_name, args, kwargs):
     return getattr(rref.local_value(), func_name)(*args, **kwargs)
 
+# convert a Future[Future[X]] to a Future[X]
+def _unwrap_future(fut: Future) -> Future:
+    res: Future = Future()
+
+    def _unwrap_level0(fut0):
+        try:
+            res.set_result(fut0.wait())
+        except BaseException as ex:
+            res.set_exception(ex)
+
+    def _unwrap_level1(fut1):
+        try:
+            fut1.wait().then(_unwrap_level0)
+        except BaseException as ex:
+            res.set_exception(ex)
+
+    fut.then(_unwrap_level1)
+    return res
+
 def _invoke_rpc(rref, rpc_api, func_name, timeout, *args, **kwargs):
-    # Since rref._get_type can potentially issue an RPC, it should respect the
-    # passed in timeout here.
-    rref_type = rref._get_type(timeout=timeout)
-
-    _invoke_func = _local_invoke
-    # Bypass ScriptModules when checking for async function attribute.
-    bypass_type = issubclass(rref_type, torch.jit.ScriptModule) or issubclass(
-        rref_type, torch._C.ScriptModule
-    )
-    if not bypass_type:
-        func = getattr(rref_type, func_name)
-        if hasattr(func, "_wrapped_async_rpc_function"):
-            _invoke_func = _local_invoke_async_execution
-
-    return rpc_api(
-        rref.owner(),
-        _invoke_func,
-        args=(rref, func_name, args, kwargs),
-        timeout=timeout
-    )
+    # this function returns either the value or a Future depending on what rpc_api is.
+    def _rref_type_cont(future, is_future):
+        if is_future:
+            rref_type = future.wait()
+        else:
+            rref_type = future
+
+        _invoke_func = _local_invoke
+        # Bypass ScriptModules when checking for async function attribute.
+        bypass_type = issubclass(rref_type, torch.jit.ScriptModule) or issubclass(
+            rref_type, torch._C.ScriptModule
+        )
+        if not bypass_type:
+            func = getattr(rref_type, func_name)
+            if hasattr(func, "_wrapped_async_rpc_function"):
+                _invoke_func = _local_invoke_async_execution
+
+        return rpc_api(
+            rref.owner(),
+            _invoke_func,
+            args=(rref, func_name, args, kwargs),
+            timeout=timeout
+        )
+
+    request_blocking = rpc_api != rpc_async
+    rref_val = rref._get_type(timeout=timeout, blocking=request_blocking)
+
+    # _get_type return type depends on how the first call is made to it
+    # so it's possible to get a Future even if you call it with `blocking=True`
+    is_future = issubclass(type(rref_val), torch._C.Future)
```

Comment:
Isn't `is_future` basically `!request_blocking`?

Reply:
Unfortunately no.

_get_type can return either `type(Future)` and `type(Class)` regardless of whether the caller requested a blocking call or not.
This behavior of `_get_type` is counter intuitive, do you think it makes sense to fix it instead of having this confusing piece of code here?

Reply:
Right but there are two cases:

1. blocking=True, in this case `_get_type` returns Class (is_future=False)
2. blocking=False, in this case `_get_type` returns Future (is_future= True)

So isn't `is_future = !request_blocking`?

Reply:
The problem is that `_get_type` behavior is surprising and doesn't follow what you'd rightfully assumed.

`_get_type()` can return `Class` or `Future` *regardless* of the `blocking` argument.
You can try this snippet: https://gist.github.com/kumpera/4a9560af769cd59aa2f955afb8b1de49
It prints:
```
[sync-first] blocking=True -> <class 'type'> (should be type)
[sync-first] blocking=False -> <class 'type'> (should be Future)
[async-first] blocking=False -> <class 'torch._C.Future'> (should be Future)
[async-first] blocking=True -> <class 'torch._C.Future'> (should be type)
```

Which makes sense once you see how `PyRRef::getRRefType` is implemented: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/py_rref.cpp#L271

It caches the value straight from the rpc call and doesn't try to compensate when the cached one came from a different `blocking` value than a just requested value.

One simplification would be to always request it with `blocking=False` and have the sync path call `wait`.